src/main/java/usi/si/seart/repository/specification/GitRepoSpecification.java

```diff
@@ -41,103 +42,28 @@ public Predicate toPredicate(
         List<Predicate> predicates = new ArrayList<>();
 
         for (Criteria criteria : criteriaList) {
-            if (criteria instanceof KeyCriteria) {
-                handle(predicates, root, criteriaBuilder, (KeyCriteria) criteria);
-            } else if (criteria instanceof KeyValueCriteria) {
-                handle(predicates, root, criteriaBuilder, (KeyValueCriteria) criteria);
-            } else if (criteria instanceof NestedKeyValueCriteria) {
-                handle(predicates, root, criteriaBuilder, (NestedKeyValueCriteria) criteria);
-            } else {
-                throw new UnsupportedOperationException(
-                        "Criteria type: ["+criteria.getClass().getCanonicalName()+"] not supported!"
-                );
-            }
+            handle(predicates, root, criteriaBuilder, criteria);
         }
 
-        return criteriaBuilder.and(predicates.toArray(new Predicate[0]));
-    }
+        query.distinct(true);
```

Comment:
Don't use `distinct`, use `groupBy` instead:
```java
query.groupBy(root.get(GitRepo_.ID));
```
The grouping should also be added conditionally since it does not need to be performed when we don't do any joins. Maybe add a separate list field with dedicated `add` method:
```java
public class GitRepoSpecification implements Specification<GitRepo> {

    // criteria list and its add method...

    Set<Expression> groupings = new LinkedHashSet<>();

    private void add(Expression expression) {
        groupings.add(expression);
    }

    /// other declarations...
```

All you would have to do afterwards is:
- Add the `groupings` collection to the query via `groupBy`
- Call the declared `add` method in the `from`